File: dwds/test/puppeteer/extension.test.js
// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

const puppeteer = require('puppeteer');

let browser;
let appUrl;
let serviceWorker;

beforeAll(async () => {
  const [_, __, ___, extensionDirArg, appUrlArg] = process.argv;
  appUrl = appUrlArg;
  const pathToExtension = require('path').join(extensionDirArg);
  browser = await puppeteer.launch({
    headless: 'chrome',
    args: [
      `--disable-extensions-except=${pathToExtension}`,
      `--load-extension=${pathToExtension}`,
      '--disable-features=DialMediaRouteProvider',
    ],
  });
  const serviceWorkerTarget = await browser.waitForTarget(
    target => target.type() === 'service_worker'
  );
  serviceWorker = await serviceWorkerTarget.worker();
});

afterAll(async () => {
  await browser.close();
});

describe('Dart Debug Extension', () => {
  test('Clicking extension icon opens Dart docs', async () => {
    // Navigate to the app tab:
    const appTab = await browser.newPage();
    await appTab.goto(appUrl, { waitUntil: ['domcontentloaded', "networkidle2"] });
    await appTab.bringToFront();

    // Click on the Dart Debug Extension icon:
    await serviceWorker.evaluate(async () => {
      const activeTabs = await chrome.tabs.query({ active: true });
      const tab = activeTabs[0];
      chrome.action.onClicked.dispatch(tab);
    });

    // Verify that the extension opened the Dart docs:
    const newTabTarget = await browser.waitForTarget(
      target => target.url() === 'https://dart.dev/'
    );
    const newTabPage = await newTabTarget.page();
    expect(newTabPage).not.toBeNull();
  });
});

File: dwds/debug_extension_mv3/puppeteer_tests/extension_test.dart
import 'package:puppeteer/puppeteer.dart';
import 'package:test/test.dart';

late Target serviceWorkerTarget;
late Browser browser;

final clickIconJs = '''
  async () => {
    const activeTabs = await chrome.tabs.query({ active: true });
    const tab = activeTabs[0];
    chrome.action.onClicked.dispatch(tab);
  }
''';

void main() async {
  group('Debug Extension Tests', () {
    setUpAll(() async {
      final pathToExtension =
          '/Users/elliottbrooks/dev/webdev/dwds/debug_extension_mv3/compiled';

      print('launching...');
      browser = await puppeteer.launch(
        headless: true,
        args: [
          '--disable-extensions-except=${pathToExtension}',
          '--load-extension=${pathToExtension}',
          '--disable-features=DialMediaRouteProvider',
        ],
      );

      print('getting sw...');

      serviceWorkerTarget = await browser
          .waitForTarget((target) => target.type == 'service_worker');
    });

    test('Can use Dart Debug Extension', () async {
      final appTab = await browser.newPage();
      await appTab.goto('https://github.com/dart-lang/webdev',
          wait: Until.all([Until.domContentLoaded, Until.networkIdle]));
      await appTab.bringToFront();

      final worker = (await serviceWorkerTarget.worker)!;
      await worker.evaluate(clickIconJs);
      // Verify that the extension opened the Dart docs:
      final newTabTarget = await browser
          .waitForTarget((target) => target.url == 'https://dart.dev/');
      final newTabPage = await newTabTarget.page;
      expect(newTabPage, isNotNull);
    });
  });
}
